### PR TITLE
Anton/crud manager

### PR DESCRIPTION
**Pull request by antonoberg**

### CODE REVIEW

File: packages/ui/src/coreModules/crud/createCrudManager/factories/selectors/index.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+const log = createLog('coreModules:crud:selectors')
+export default function createResourceSelectors(
+  { resourceSpecification } = {}
+) {
+  const { resource } = resourceSpecification
+  if (!resource) {
+    throw new Error('resource required')
+  }
+  const resourceSelectors = {}
```
Comment (fredrikolovsson) [inline packages/ui/src/coreModules/crud/createCrudManager/factories/selectors/index.js]: this pattern seems unnecessarily complicated here. 

suggestion, which is more similar to how we tend to write selectors elsewhere:

```javascript
const getLocalState = state => {
    return state[MODULE_NAME] && state[MODULE_NAME][RESOURCES_NAMESPACE]
  }

  const getLocalResourceState = localState => {
    return localState[resource]
  }

  const getItemsObject = localState => {
    const localResourceState = getLocalResourceState(localState)
    return localResourceState.items
  }

  const getAll = createSelector(getItemsObject, items => {
    return Object.values(items)
  })

  const getOne = (localState, id) => {
    const items = getItemsObject(localState)
    return items[id] || null
  }

  const resourceSelectors = {
    getAll,
    getItemsObject,
    getLocalResourceState,
    getLocalState,
    getOne,
  }

  Object.keys(resourceSelectors).forEach(selectorName => {
    log.info(`Adding selector ${resource}.${selectorName}`)
  })
```

File: packages/ui/src/coreModules/crud/higherOrderComponents/createEnsureAllItemsFetched.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import config from 'config'
+import actionCreators from '../actionCreators'
+import {
+  actionCreators as keyObjectActionCreators,
+  globalSelectors as keyObjectGlobalSelectors,
+} from '../keyObjectModule'
+
+const createEnsureAllItemsFetched = ({
+  resource,
+  relationships = ['parent'],
```
Comment (fredrikolovsson) [inline packages/ui/src/coreModules/crud/higherOrderComponents/createEnsureAllItemsFetched.js]: I think the default should be no relationships. not all resources have a parent relationship

File: packages/common/src/jsonApiClient/index.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
+const { createApiClient } = require('../apiClient')
+
+module.exports = function createJsonApiClient({
+  apiConfigInput,
+  createEndpoint,
+}) {
+  const apiClient = createApiClient(apiConfigInput)
+  const enpointConfigCache = {}
+  const call = (input, userInput) => {
+    let enpointConfig
```
Comment (fredrikolovsson) [inline packages/common/src/jsonApiClient/index.js]: => endpointConfig (think it's worth changing so it shows up when searching in the repo for `endpointConfig`

File: packages/ui/src/coreModules/crud/createCrudManager/factories/actionHandlers/updateStateWithOneFactory.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      }. Updating state from action: `,
+      action
+    )
+
+    const { id } = action.payload
+    const updatePath = dep.createItemUpdatePath({
+      id,
+    })
+
+    const newState = {
```
[the remark this answers is not in the document]
Comment (antonoberg) [reply packages/ui/src/coreModules/crud/createCrudManager/factories/actionHandlers/updateStateWithOneFactory.js]: The point is to ensure that the state has got items. However this should be set in the reducer as initial state instead. Will fix

File: packages/ui/src/apps/collectionsUi/viewModules/app/Async.js [hunk 5]
```diff
@@ -2,6 +2,9 @@ import { createAsyncView } from 'coreModules/bootstrap/higherOrderComponents'
 import { MODULE_NAME } from './constants'
 
 export default createAsyncView({
+  modules: () => {
+    return [import('coreModules/crud')]
```
[the remark this answers is not in the document]
Comment (antonoberg) [reply packages/ui/src/apps/collectionsUi/viewModules/app/Async.js]: Its not needed when you are not logged in and i want the data to be removed as the user loggs out.

File: packages/ui/src/coreModules/crud/createCrudManager/index.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import dux from './factories/dux'
+import createSpecification from './factories/specification'
+
+export default function createCrudManager(config) {
```
[the remark this answers is not in the document]
Comment (antonoberg) [reply packages/ui/src/coreModules/crud/createCrudManager/index.js]: Yes

File: packages/ui/src/coreModules/crud/createCrudManager/factories/dux/index.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+const log = createLog('coreModules:crud:dux')
+
+export default function dux(specification) {
+  const selectors = {}
+  const globalSelectors = {}
+  const actionTypes = {}
+  const actionHandlers = {}
+  const actionCreators = {}
+  const resourceReducers = {}
+  iterateResources(specification, resourceSpecification => {
```
[the remark this answers is not in the document]
Comment (antonoberg) [reply packages/ui/src/coreModules/crud/createCrudManager/factories/dux/index.js]: True. will fix

File: packages/ui/src/coreModules/crud/createCrudManager/factories/actionCreators/updateFactory.test.js [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+import createApiActionTypeFailTest from './utilities/test/createApiActionTypeFailTest'
+import createApiActionTypeRequestTest from './utilities/test/createApiActionTypeRequestTest'
+import createApiActionTypeSuccessTest from './utilities/test/createApiActionTypeSuccessTest'
+import createApiClientTest from './utilities/test/createApiClientTest'
+import updateFactory, { dep } from './updateFactory'
+
+describe('coreModules/crud/createCrudManager/factories/actionCreators/updateFactory', () => {
+  describe('factory', () => {
+    const actionCreatorFactoryInput = {
+      operationId: 'createPhysicalObject',
```
[the remark this answers is not in the document]
Comment (antonoberg) [reply packages/ui/src/coreModules/crud/createCrudManager/factories/actionCreators/updateFactory.test.js]: yes

File: packages/ui/src/coreModules/crud/createCrudManager/factories/actionHandlers/updateStateWithManyFactory.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        })
+
+        if (!updatePath) {
+          return newState
+        }
+
+        return dep.assign(newState, updatePath, item)
+      },
+      {
+        ...state,
```
[the remark this answers is not in the document]
Comment (antonoberg) [reply packages/ui/src/coreModules/crud/createCrudManager/factories/actionHandlers/updateStateWithManyFactory.js]: same as above. will fix

File: packages/ui/src/coreModules/crud/createCrudManager/factories/selectors/index.js (hunk 1 above)
Comment (antonoberg) [reply packages/ui/src/coreModules/crud/createCrudManager/factories/selectors/index.js]: sure will change

File: packages/ui/src/coreModules/crud/higherOrderComponents/createEnsureAllItemsFetched.js (hunk 2 above)
Comment (antonoberg) [reply packages/ui/src/coreModules/crud/higherOrderComponents/createEnsureAllItemsFetched.js]: true -> will change

File: packages/common/src/jsonApiClient/index.js (hunk 3 above)
Comment (antonoberg) [reply packages/common/src/jsonApiClient/index.js]: haha true

File: packages/ui/src/coreModules/crud/higherOrderComponents/createEnsureAllItemsFetched.js (hunk 2 above)
Comment (antonoberg) [reply packages/ui/src/coreModules/crud/higherOrderComponents/createEnsureAllItemsFetched.js]: fc5acd8

File: packages/ui/src/coreModules/crud/createCrudManager/factories/selectors/index.js (hunk 1 above)
Comment (antonoberg) [reply packages/ui/src/coreModules/crud/createCrudManager/factories/selectors/index.js]: bfa5b51

File: packages/ui/src/coreModules/crud/createCrudManager/factories/actionHandlers/updateStateWithOneFactory.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (antonoberg) [reply packages/ui/src/coreModules/crud/createCrudManager/factories/actionHandlers/updateStateWithOneFactory.js]: 20dd995

File: packages/ui/src/coreModules/crud/createCrudManager/factories/actionHandlers/updateStateWithManyFactory.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (antonoberg) [reply packages/ui/src/coreModules/crud/createCrudManager/factories/actionHandlers/updateStateWithManyFactory.js]: 20dd995

File: packages/ui/src/coreModules/crud/createCrudManager/factories/actionCreators/updateFactory.test.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (antonoberg) [reply packages/ui/src/coreModules/crud/createCrudManager/factories/actionCreators/updateFactory.test.js]: 62693e7

File: packages/ui/src/coreModules/crud/createCrudManager/factories/dux/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (antonoberg) [reply packages/ui/src/coreModules/crud/createCrudManager/factories/dux/index.js]: 93ee21f

File: packages/ui/src/coreModules/crud/createCrudManager/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (antonoberg) [reply packages/ui/src/coreModules/crud/createCrudManager/index.js]: 8f0d863

**Pull request merged**